scripts/taxonomies/lint_taxonomy.pl

```diff
@@ -0,0 +1,507 @@
+	# iterator
+	return sub {
+		while (my $line = <$fd>) {
+			$last_line = $line;
+			return $line;
+		}
+		# be sure to end with a blank line
+		unless ($last_line =~ /^\s*$/) {
+			$last_line = "\n";    # make next call terminate
+			return "\n";
+		}
+		# end of iteratorsanitize_taxonomy_line
+		return;
+	}
+}
+
+# iter over the taxonomy entry by entry
+# return a ref to a hash map with entry infos
+sub iter_taxonomy_entries ($lines_iter) {
+	my $line_num = 0;    # this is global
+	return sub {
+		# re-init at start and after returning each entry
+		my @parents = ();    # lines defining parents
+		my $entry_id_line = undef;    # line defining entry id, we don't want to change it's position
+		my %entries = ();    # lines defining synonyms
+		my %props = ();    # lines defining properties
+		my @original_lines = ();
+		# non meaningful lines above a meaningful line (entry, parent or porperty)
+		my @previous_lines = ();
+		my @errors = ();
+		my $entry_start_line = $line_num + 1;
+		while (my $line = $lines_iter->()) {
+			$line_num += 1;
+			push @original_lines, $line;    # collect lines for comparison
+
+			# blank line means we are changing entry, so let's return collected data
+			if ($line =~ /^\s*$/) {
+				my $entry = {
+					parents => \@parents,
+					entry_id_line => $entry_id_line,
+					entries => \%entries,
+					props => \%props,
+					original_lines => \@original_lines,
+					tail_lines => \@previous_lines,
+					start_line => $entry_start_line,
+					end_line => $line_num,
+					errors => \@errors,
+				};
+				add_entry_id($entry, \@errors);
+				# return $entry
+				return $entry;
+			}
+			# parents
+			elsif ($line =~ /^</) {
+				# if we already have an entry id, parent are not at the good position,
+				# and it might mean two entries where merged inadvertantly (eg. by an auto merge)
+				if ($entry_id_line) {
+					push(
+						@errors,
+						{
+							severity => "Error",
+							type => "Correctness",
+							line => $line_num,
+							message => (
+									  "Parent in the middle of an entry, might mean erroneous merge of two entries:\n"
+									. "- $line"
+							)
+						}
+					);
+				}
+				push @parents, {line => $line, previous => [@previous_lines], line_num => $line_num};
+				@previous_lines = ();
+			}
+			# synonym
+			elsif ($line =~ /^(\w+):[^:]*(,.*)*$/) {
+				if (!defined $entry_id_line) {
+					$entry_id_line = {line => $line, previous => [@previous_lines], lc => $1,, line_num => $line_num};
+				}
+				else {
+					my $lc = $1;
+					if ((defined $entries{$lc}) || ($entry_id_line->{lc} eq $lc)) {
+						my $previous_lc_line;
+						if (defined $entries{$lc}) {
+							$previous_lc_line = $entries{$lc}{line};
+						}
+						else {
+							$previous_lc_line = $entry_id_line->{line};
+						}
+
+						push @errors,
+							{
+							severity => "Error",
+							type => "Correctness",
+							line => $line_num,
+							message => ("duplicate synonym for $lc:\n" . "- $previous_lc_line" . "- $line")
```

Comment:
```suggestion
							message => ("duplicate line for $lc:\n" . "- $previous_lc_line" . "- $line")
```

Reply:
> if we change that, we should change the code variables etc. that use "duplicate_synonym"

No, the type, is only used in Tags.pm, for now.